### PR TITLE
[BEAM-14347] Add function for simple function registration

### PR DESCRIPTION
**Pull request by damccorm**
This adds a basic function for non-structural DoFn registration. This doesn't actually need to be generic, and it doesn't have to be aware of its parameters (we could just accept an interface). I chose to do it this way because (1) it gives us greater type safety and (2) it matches the other registration functions

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @lostluck here's the simple function registration you suggested. Arguably we don't need it, but it was simple enough and gives us full consistency

**Comment by lostluck** [comment]
> R: @lostluck here's the simple function registration you suggested. Arguably we don't need it, but it was simple enough and gives us full consistency

We do need it for the same reason. Please add a benchmark for wrapping and invoking a simple function caller, and you'll see.

The stats combiners are simple binary functions eg [for minIntFn](https://github.com/apache/beam/blob/243128a8fc52798e1b58b0cf1a271d95ee7aa241/sdks/go/pkg/beam/transforms/stats/min_switch.go#L56)
The current code generator [generates registrations for them all](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/transforms/stats/stats.shims.go#L52) and their [factory wrappers](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/transforms/stats/stats.shims.go#L81) specifically for the [caller invoker](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/transforms/stats/stats.shims.go#L261).  

The various struct wrappers happened 2nd originally, and piggy backed on the work to make functions faster (partly because that was faster than an alternative method receiver passing approach).

File: sdks/go/pkg/beam/register/register.tmpl [hunk 1]
```diff
@@ -194,11 +194,18 @@ func registerDoFn{{$processElementIn}}x{{$processElementOut}}StructWrappersAndFu
...
-// DoFn{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut true)}} registers your DoFn to optimize execution at runtime.
+// DoFn{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut true)}} registers your structural DoFn to optimize execution at runtime.
 {{if (or $processElementIn $processElementOut)}}// DoFn input and output parameter types should be provided in order as the generic constraints.
 {{end}}func DoFn{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut true)}}(doFn genericDoFn{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut false)}}) {
     registerDoFnTypes(doFn)
     registerDoFn{{$processElementIn}}x{{$processElementOut}}StructWrappersAndFuncs{{(genericTypingRepresentation $processElementIn $processElementOut false)}}(doFn)
+}
+
+// Function{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut true)}} registers your non-structural DoFn to optimize execution at runtime.
```
Comment (lostluck) [inline sdks/go/pkg/beam/register/register.tmpl:204]:
```suggestion
// Function{{$processElementIn}}x{{$processElementOut}}{{(genericTypingRepresentation $processElementIn $processElementOut true)}} registers your functional DoFn to optimize execution at runtime.
```

**Comment by damccorm** [comment]
> We do need it for the same reason. Please add a benchmark for wrapping and invoking a simple function caller, and you'll see.

Oh yeah, my point isn't that it isn't much better in terms of performance, just that its not a huge usability win vs the user manually calling these themselves. The performance bump is evident from the existing benchmarks on `register.DoFnXxY` which are doing the same thing, I'm pretty happy to add the extra benchmark though, and will have it shortly.

I should've been more clear in my original comment that I **do** still think this is a good idea, I just think it is less useful because its not too hard for users to do the right thing themselves already.

File: sdks/go/pkg/beam/register/register.tmpl (hunk 1 above)
Comment (damccorm) [reply sdks/go/pkg/beam/register/register.tmpl:204]: I won't commit it directly since it needs to match the registered version, but agreed this is better

**Comment by lostluck** [comment]
Oh fair!
But, as written, you aren't generating callers for the user, just the simple type and func registrations. The generic callers are where the execution time performance comes from, and they don't exist in your template.

**Comment by damccorm** [comment]
> But, as written, you aren't generating callers for the user, just the simple type and func registrations. The generic callers are where the execution time performance comes from, and they don't exist in your template.

Ah, you're right! Good callout, I'll fix it

**Comment by damccorm** [comment]
Fixed with benchmarks! Thanks for calling this out, I'd gotten confused on what actually gets generated for these ones (which is also why I thought it was an easy one for users to do themselves)

File: sdks/go/pkg/beam/register/register_test.go [hunk 2]
```diff
@@ -818,25 +865,37 @@ func BenchmarkMethodCalls(b *testing.B) {
...
+		{"EncodeMultiEdge_Unoptimized", func() { aME, _ = graphx.EncodeMultiEdge(&me) }, func() { /*No op*/ }},                                                                                             // Used in graph serialization at execution time
+		{"MakeFunc_FunctionalDoFn_Unoptimized", func() { aFunc = reflectx.MakeFunc(addCustomType2) }, func() { /*No op*/ }},                                                                                // Used in graph deserialization
+		{"MakeFunc_FunctionalDoFn.Call_Unoptimized", func() { aFnCall2 = aFunc.Call(funcIn2)[0].(CustomType2) }, func() { /*No op*/ }},                                                                     // Used to call the function repeatedly
+		{"MakeFunc_FunctionalDoFn.Call1x1_Unoptimized", func() { aFnCall2 = aFunc2x1.Call2x1(CustomType2{val2: 4}, CustomType2{val2: 3}).(CustomType2) }, func() { aFunc2x1 = reflectx.ToFunc2x1(aFunc) }}, // Used to call the function repeatedly
```
Comment (lostluck) [inline sdks/go/pkg/beam/register/register_test.go:875]: Typo: 2x1 here and below

Comment (lostluck) [reply sdks/go/pkg/beam/register/register_test.go:875]: I guess just here.

File: sdks/go/pkg/beam/register/register_test.go [hunk 3]
```diff
@@ -818,25 +865,37 @@ func BenchmarkMethodCalls(b *testing.B) {
...
+		{"EncodeMultiEdge_GeneratedShims", func() { aME, err = graphx.EncodeMultiEdge(&me) }, func() { GeneratedOptimizationCalls() }},                                                                                                   // Used in graph serialization at execution time
+		{"MakeFunc_FunctionalDoFn_Unoptimized", func() { aFunc = reflectx.MakeFunc(addCustomType2) }, func() { GeneratedOptimizationCalls() }},                                                                                           // Used in graph deserialization
+		{"MakeFunc_FunctionalDoFn.Call_Unoptimized", func() { aFnCall2 = aFunc.Call(funcIn2)[0].(CustomType2) }, func() { GeneratedOptimizationCalls() }},                                                                                // Used to call the function repeatedly
+		{"MakeFunc_FunctionalDoFn.Call1x1_Unoptimized", func() { aFnCall2 = aFunc2x1.Call2x1(CustomType2{val2: 4}, CustomType2{val2: 3}).(CustomType2) }, func() { GeneratedOptimizationCalls(); aFunc2x1 = reflectx.ToFunc2x1(aFunc) }}, // Used to call the function repeatedly
```
Comment (lostluck) [inline sdks/go/pkg/beam/register/register_test.go:898]: And this should be the "generated" not optimized. (also 2x1)

Comment (damccorm) [reply sdks/go/pkg/beam/register/register_test.go]: Oop - fixed both, thanks!

**Review by lostluck — APPROVED** [review]
LGTM

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Pull request merged**